### PR TITLE
LPD-32882 Apply new validator to Returns for checking custom permissions

### PR DESCRIPTION
**Pull request by ghost**
https://liferay.atlassian.net/browse/LPD-32882

### CODE REVIEW

**Comment by CrescenzoRegaLF** [comment]
ci:test:sf

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnItemPermissionValidationRuleEngineImpl.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Crescenzo Rega
+ */
+@Component(service = ObjectValidationRuleEngine.class)
+public class CommerceReturnItemPermissionValidationRuleEngineImpl
+	implements ObjectDefinitionScoped, ObjectValidationRuleEngine {
```
Comment (dsitu) [inline modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnItemPermissionValidationRuleEngineImpl.java:38]: I think we can just make this extend BaseObjectValidationRuleEngineImpl so we can follow the pattern that bchan set

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnPermissionValidationRuleEngineImpl.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Crescenzo Rega
+ */
+@Component(service = ObjectValidationRuleEngine.class)
+public class CommerceReturnPermissionValidationRuleEngineImpl
+	implements ObjectDefinitionScoped, ObjectValidationRuleEngine {
```
Comment (dsitu) [inline modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnPermissionValidationRuleEngineImpl.java:34]: I think we can just make this extend BaseObjectValidationRuleEngineImpl so we can follow the pattern that bchan set

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnItemPermissionValidationRuleEngineImpl.java (hunk 1 above)
Comment (CrescenzoRegaLF) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnItemPermissionValidationRuleEngineImpl.java:38]: Validators that extend the `BaseObjectValidationRuleEngineImpl` are validators that validate a specific field, these new validators instead do not validate a field but prevent the modification of some fields if the Return status is different from DRAFT and if the user does not have the MANAGE_RETURNS permission

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnPermissionValidationRuleEngineImpl.java (hunk 2 above)
Comment (CrescenzoRegaLF) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnPermissionValidationRuleEngineImpl.java:34]: Validators that extend the `BaseObjectValidationRuleEngineImpl` are validators that validate a specific field, these new validators instead do not validate a field but prevent the modification of some fields if the Return status is different from DRAFT and if the user does not have the MANAGE_RETURNS permission

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnItemPermissionValidationRuleEngineImpl.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+import java.util.Objects;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Crescenzo Rega
+ */
+@Component(service = ObjectValidationRuleEngine.class)
+public class CommerceReturnItemPermissionValidationRuleEngineImpl
```
[the remark this answers is not in the document]
Comment (CrescenzoRegaLF) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnItemPermissionValidationRuleEngineImpl.java:37]: Yes I am aware that the name is incorrect, but I still have to figure out what to name the new validators

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnPermissionValidationRuleEngineImpl.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.Objects;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Crescenzo Rega
+ */
+@Component(service = ObjectValidationRuleEngine.class)
+public class CommerceReturnPermissionValidationRuleEngineImpl
```
[the remark this answers is not in the document]
Comment (CrescenzoRegaLF) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnPermissionValidationRuleEngineImpl.java:33]: Yes I am aware that the name is incorrect, but I still have to figure out what to name the new validators

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnItemPermissionValidationRuleEngineImpl.java (hunk 1 above)
Comment (dsitu) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/CommerceReturnItemPermissionValidationRuleEngineImpl.java]: We can look into creating something like BasePermissionObjectValidationRuleEngineImpl, if that makes more sense

**Comment by CrescenzoRegaLF** [comment]
ci:test:sf

**Comment by CrescenzoRegaLF** [comment]
ci:test:sf

**Comment by CrescenzoRegaLF** [comment]
ci:test:relevant

**Comment by CrescenzoRegaLF** [comment]
ci:test:sf

**Comment by CrescenzoRegaLF** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/BasePermissionObjectValidationRuleEngineImpl.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+	@Override
+	public String getLabel(Locale locale) {
+		return LanguageUtil.get(
+			locale,
+			TextFormatter.format(
+				getObjectDefinitionName() + "-permission", TextFormatter.K));
+	}
+
+	protected boolean equals(
```
Comment (dsitu) [inline modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/BasePermissionObjectValidationRuleEngineImpl.java:56]:
```suggestion
	protected boolean isEqual(
```

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/BasePermissionObjectValidationRuleEngineImpl.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		Map<String, Object> map1, Map<String, Object> map2, String... keys) {
+
+		for (String key : keys) {
+			if (!Objects.equals(map1.get(key), map2.get(key))) {
+				return false;
+			}
+		}
+
+		return true;
+	}
```
Comment (dsitu) [inline modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/object/validation/rule/BasePermissionObjectValidationRuleEngineImpl.java:66]:
```suggestion
	protected boolean hasChanges(
		Map<String, Object> map1, Map<String, Object> map2, String... keys) {

		for (String key : keys) {
			if (!Objects.equals(map1.get(key), map2.get(key))) {
				return true;
			}
		}

		return false;
	}
```

**Comment by CrescenzoRegaLF** [comment]
ci:test:sf

**Comment by CrescenzoRegaLF** [comment]
ci:test:relevant

**Comment by CrescenzoRegaLF** [comment]
ci:stop

**Comment by CrescenzoRegaLF** [comment]
ci:test:relevant

**Comment by luca-pellizzon** [comment]
ci:stop

**Comment by luca-pellizzon** [comment]
ci:test:relevant

**Comment by dsitu** [comment]
Closing opened up new PR: https://github.com/liferay-commerce/liferay-portal/pull/5248, changes https://github.com/liferay-commerce/liferay-portal/pull/5248/commits/ed2d54fe36fbea72cb85827c28b560ae0f26c30a, https://github.com/liferay-commerce/liferay-portal/pull/5248/commits/c80757a08c8f3ace30e85c9505b47e6f9fc097a5

**Pull request closed without merging**